xls/contrib/integrator/ir_integrator.cc

```diff
@@ -194,6 +244,168 @@ absl::StatusOr<Node*> IntegrationFunction::InsertNode(const Node* to_insert) {
+    merge_result.other_added_nodes.insert(merged);
+    return merge_result;
+  } else {
+    // TODO(jbaileyhandle): Add logic for nodes that are not identical but
+    // may still be merged e.g. different bitwidths for bitwise ops.
+    switch (node_a->op()) {
+      default:
+        break;
+    }
+  }
+
+  MergeNodesBackendResult merge_result;
+  merge_result.can_merge = false;
+  return merge_result;
+}
+
+absl::StatusOr<std::optional<float>> IntegrationFunction::GetMergeNodesCost(
+    const Node* node_a, const Node* node_b) {
+  XLS_ASSIGN_OR_RETURN(MergeNodesBackendResult merge_result,
+                       MergeNodesBackend(node_a, node_b));
+  // Can't merge nodes.
+  if (!merge_result.can_merge) {
+    return absl::nullopt;
+  }
+
+  // Score.
+  float cost = 0.0;
+  auto tabulate_node_elimination_cost = [this, &cost](const Node* node) {
+    if (IntegrationFunctionOwnsNode(node)) {
+      cost -= GetNodeCost(node);
+    }
+  };
+  tabulate_node_elimination_cost(node_a);
+  tabulate_node_elimination_cost(node_b);
+  for (Node* new_node : merge_result.added_muxes) {
+    cost += GetNodeCost(new_node);
+  }
+  for (Node* new_node : merge_result.other_added_nodes) {
+    cost += GetNodeCost(new_node);
+  }
+
+  // Cleanup.
+  int64 nodes_eliminated = 0;
+  int64 num_other_nodes_to_remove = merge_result.other_added_nodes.size();
+  while (nodes_eliminated < num_other_nodes_to_remove) {
+    int64 initial_eliminated = nodes_eliminated;
+    for (Node* new_node : merge_result.other_added_nodes) {
+      if (new_node->users().empty()) {
+        merge_result.other_added_nodes.erase(new_node);
+        XLS_RETURN_IF_ERROR(function()->RemoveNode(new_node));
+        ++nodes_eliminated;
+        break;
+      }
+    }
+    XLS_RET_CHECK_GT(nodes_eliminated, initial_eliminated);
+  }
+  for (Node* new_node : merge_result.added_muxes) {
+    XLS_RETURN_IF_ERROR(DeUnifyIntegrationNodes(new_node));
+  }
+
+  return cost;
+}
+
+float IntegrationFunction::GetNodeCost(const Node* node) const {
+  // TODO: Actual estimate.
+  switch (node->op()) {
+    case Op::kArray:
+    case Op::kConcat:
+    case Op::kBitSlice:
+    case Op::kIdentity:
+    case Op::kParam:
+    case Op::kReverse:
+    case Op::kSignExt:
+    case Op::kTuple:
+    case Op::kTupleIndex:
+    case Op::kZeroExt:
+      return 0.0;
+      break;
+    default:
+      return 1.0;
+      break;
+  }
+}
+
+absl::StatusOr<std::vector<Node*>> IntegrationFunction::MergeNodes(
+    Node* node_a, Node* node_b) {
+  XLS_ASSIGN_OR_RETURN(MergeNodesBackendResult merge_result,
+                       MergeNodesBackend(node_a, node_b));
+  // Can't merge nodes.
+  XLS_RET_CHECK(merge_result.can_merge);
+  XLS_RET_CHECK(merge_result.target_a != nullptr);
+  XLS_RET_CHECK(merge_result.target_b != nullptr);
+
+  // Commit changes.
+  auto commit_new_node = [this](Node* original, Node* target) {
+    auto map_status = SetNodeMapping(original, target);
```

Comment:
You can use macro here XLS_ASSIGN_OR_RETURN()

I believe you'll also have to declare the return type of the lambda:

auto commit_new_node = [...](...) -> absl::Status {

Reply:
Done.